designs/0034-jacobian.md

```diff
@@ -0,0 +1,381 @@
+This is very nice for both parties since it is both of their preferred spaces to work in.
+
+Most of the time this encapsulation is very good. 
+The main issue is that Stan users either have to write code that either only uses Stan's built-in transforms to stay in the constrained space or works directly on the unconstrained space. 
+For instance, from a comment [2] by @betanalpha (slightly commented and fleshed out)
+
+-----------
+
+> It’s only in the context of trying to assign a density on the output value that a Jacobian is needed, so the proper encapsulation is
+
+```stan
+functions {
+ real f(real x) {
+   // transform function
+ }
+ real jacobian(real x) {
+   // jacobian determinant calculation
+ }
+ real prob(real x) {
+   // lpdf calculation
+ }
+}
+parameters {
+ real theta;
+}
+transformed parameters {
+ real eta = f(theta);
+}
+model {
+ // In a pure sense, neither line is well defined on it's
+ //  own, but the combinations are well defined.
+ eta ~ prob(values);
+ target += jacobian(theta);
+}
+```
+
+-----------
+
+Coding transforms directly across multiple blocks (parameter, transformed parameter, model) is clunky and error prone. 
+And when users want to add a constraint to the Stan language they have to touch the math library which is rather complicated. 
+Not just by C++ itself, but the essentially small DSL we use to describe autodiff in the Stan math library.
+
+Having a new block for user-defined constraints allows users to encapsulate and reuse their transforms as units.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The following section is a draft of the docs for the new constraint and `jacobian` keyword.
+
+## Constraint Block
+
+### Overview
+
+The constraints block was included in Stan 2.XX to provide users with the ability to define custom constraints on parameters, transformed parameters, and data. 
+This block is situated immediately after the `functions` block and before the `data` block in a Stan program. 
+The primary purpose of the `constraints` block is to allow users to define custom constraints that can be used in the same way as built-in constraints.
+
+### Syntax
+
+A constraints block is defined by the keyword `constraints`. 
+Inside of the `constraints` block each constraint is its own block with a given `constraint_name` and the functions needed to define the constraint.
+Inside the `constraints` block, users can define one or more constraint types. Each constraint type is encapsulated within its own set of curly braces and is identified by a unique name.
+
+Each constraint type must specify a set of functions that define the transformation between unconstrained and constrained spaces, validate data, and increment the log of the absolute value of the determinant of the Jacobian. 
```

Comment:
I think the "must" here needs to be revisited based on the discussion above. I think in the intro you could just drop it and say "Each constraint type specifies" and elaborate later on the requirements?

Reply:
Ah, I see you do get more specific in the section "Errors". I think you might want to call that "Requirements for a custom constraint" or similar.